### PR TITLE
cpu-stress Chaos Experiment

### LINKED ISSUE #517: Add CPU Stress Chaos Experiment
##  FEATURE REQUEST
Add m-agent based cpu-stress chaos experiment for chaos testing the CPU under load for Linux based machines

### PR DESCRIPTION
**Pull request by neelanjan00**
## Why We Need This PR
Adds m-agent based cpu-stress experiment which can stress the CPU of target machines
- It uses stress-ng tool to stress the CPU

## Which Issue This PR Fixes 
fixes #517

## Special notes for your reviewer
Experiment execution logs:
```
INFO[2022-05-11T20:31:15+05:30] Experiment Name: cpu-stress                  
INFO[2022-05-11T20:31:15+05:30] [PreReq]: Getting the ENV for the  experiment 
INFO[2022-05-11T20:31:15+05:30] [PreReq]: Updating the chaos result of cpu-stress experiment (SOT) 
INFO[2022-05-11T20:31:17+05:30] [Info]: The stress process parameters are as follows  Load Percentage=50 CPUs=1
INFO[2022-05-11T20:31:17+05:30] [Status]: Connecting to the agents           
INFO[2022-05-11T20:31:18+05:30] [Status]: Verify that stress-ng is available in the target machine 
INFO[2022-05-11T20:31:19+05:30] [Chaos]: Injecting CPU stress for 34.136.237.160:41365 agent endpoint 
INFO[2022-05-11T20:31:20+05:30] [Chaos]: CPU stress chaos injected successfully in 34.136.237.160:41365 agent endpoint 
INFO[2022-05-11T20:31:20+05:30] [Wait]: Waiting for chaos interval of 30s    
INFO[2022-05-11T20:31:50+05:30] [Chaos]: Reverting CPU stress for 34.136.237.160:41365 agent endpoint 
INFO[2022-05-11T20:31:51+05:30] [Status]: Disconnecting the agents           
INFO[2022-05-11T20:31:51+05:30] [The End]: Updating the chaos result of cpu-stress experiment (EOT)
```

**Checklist:**
-   [x] Fixes #<issue number>
-   [ ] PR messages has document related information
-   [ ] Labelled this PR & related issue with `breaking-changes` tag
-   [ ] PR messages has breaking changes related information
-   [ ] Labelled this PR & related issue with `requires-upgrade` tag
-   [ ] PR messages has upgrade related information
-   [ ] Commit has unit tests
-   [ ] Commit has integration tests
-   [x] E2E run Required for the changes

## Dependencied
- https://github.com/litmuschaos/litmus-go/pull/525

### CODE REVIEW

**Comment by avaakash** [comment]
Should we rebase the m-agent branch with the master branch, and then take this PR?
cc: @uditgaurav @ksatchit

**Comment by uditgaurav** [comment]
Rebasing the m-agent branch gives conflicts - we need to resolve them!

File: chaoslib/litmus/cpu-stress/lib/cpu-stress.go [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+
+					return errors.Errorf("error occured while injecting CPU stress chaos for %s agent endpoint, err: %s", agentEndpointList[i], agentError)
+				}
+
+				return errors.Errorf("unintelligible feedback received from agent: %s", feedback)
+			}
+
+			underChaosEndpoints = append(underChaosEndpoints, i)
+
+			common.SetTargets(agentEndpointList[i], "injected", "CPU", chaosDetails)
```
Comment (uditgaurav) [inline chaoslib/litmus/cpu-stress/lib/cpu-stress.go:129]:
```suggestion
			common.SetTargets(agentEndpointList[i], "injected", "pod", chaosDetails)
```

File: chaoslib/litmus/cpu-stress/lib/cpu-stress.go [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+
+					return errors.Errorf("error occured while reverting CPU stress chaos for %s agent endpoint, err: %s", agentEndpointList[i], agentError)
+				}
+
+				return errors.Errorf("unintelligible feedback received from agent: %s", feedback)
+			}
+
+			underChaosEndpoints = underChaosEndpoints[:len(underChaosEndpoints)-1]
+
+			common.SetTargets(agentEndpointList[i], "reverted", "CPU", chaosDetails)
```
Comment (uditgaurav) [inline chaoslib/litmus/cpu-stress/lib/cpu-stress.go:170]:
```suggestion
			common.SetTargets(agentEndpointList[i], "reverted", "pod", chaosDetails)
```

File: chaoslib/litmus/cpu-stress/lib/cpu-stress.go [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+	// Catch and relay certain signal(s) to abort channel.
+	signal.Notify(abort, os.Interrupt, syscall.SIGTERM)
+
+	// waiting for the ramp time before chaos injection
+	if experimentsDetails.RampTime != 0 {
+		log.Infof("[Ramp]: Waiting for the %vs ramp time before injecting chaos", experimentsDetails.RampTime)
+		common.WaitForDuration(experimentsDetails.RampTime)
+	}
+
+	agentEndpointList := strings.Split(experimentsDetails.AgentEndpoints, ",")
```
Comment (uditgaurav) [inline chaoslib/litmus/cpu-stress/lib/cpu-stress.go:54]: What if the agent endpoints is empty? we can handle that case.

**Comment by uditgaurav** [comment]
The build should pass in the CI @neelanjan00 !

File: chaoslib/litmus/cpu-stress/lib/cpu-stress.go (hunk 1 above)
Comment (neelanjan00) [reply chaoslib/litmus/cpu-stress/lib/cpu-stress.go:129]: We're targetting the CPU here, not a pod.

File: chaoslib/litmus/cpu-stress/lib/cpu-stress.go (hunk 3 above)
Comment (neelanjan00) [reply chaoslib/litmus/cpu-stress/lib/cpu-stress.go:54]: That condition is being checked in the function at Line #76, experiments/os/cpu-stress/experiment/cpu-stress.go

File: chaoslib/litmus/cpu-stress/lib/cpu-stress.go (hunk 1 above)
Comment (uditgaurav) [reply chaoslib/litmus/cpu-stress/lib/cpu-stress.go:129]: We should call it VM?

Comment (neelanjan00) [reply chaoslib/litmus/cpu-stress/lib/cpu-stress.go:129]: If we use VM then it will be hard to differentiate it from other m-agent based experiments...

Comment (uditgaurav) [reply chaoslib/litmus/cpu-stress/lib/cpu-stress.go:129]: I think that should be fine right, we have different pod level experiments like network, stress all having targets as pod.

**Review by uditgaurav — APPROVED** [review]

**Review by avaakash — APPROVED** [review]

**Review by ksatchit — APPROVED** [review]

**Pull request merged**